react-common/components/controls/TeachingBubble.tsx

```diff
@@ -0,0 +1,291 @@
+        left = (targetBounds.width - bubbleBounds.width) / 2 + targetBounds.left;
+    }
+
+    const updatePosition = (element: HTMLElement, top: string, left: string) => {
+        element.style.top = top;
+        element.style.left = left;
+    }
+
+    const closeLabel = lf("Close");
+    const backLabel = lf("Back");
+    const nextLabel = lf("Next");
+    const confirmLabel = lf("Got it");
+    const finishLabel = lf("Finish");
+    const hasPrevious = stepNumber > 1;
+    const hasNext = stepNumber < totalSteps;
+    const hasSteps = totalSteps > 1;
+
+    const classes = classList(
+        "teaching-bubble-container",
+        className
+    );
+
+    return ReactDOM.createPortal(<FocusTrap className={classes} onEscape={onClose}>
+        <div className="teaching-bubble-cutout"></div>
+        <div className="teaching-bubble-arrow"></div>
+        <div id={id}
+            className="teaching-bubble"
+            role={role || "dialog"}
+            aria-hidden={ariaHidden}
+            aria-label={ariaLabel}
+            aria-describedby={ariaDescribedBy}
+            aria-labelledby="teaching-bubble-title">
+            <Button
+                className="teaching-bubble-close"
+                onClick={onClose}
+                title={closeLabel}
+                ariaLabel={closeLabel}
+                rightIcon="fas fa-times-circle"
+            />
+            <div className="teaching-bubble-content">
+                <strong>{targetContent.title}</strong>
+                <p>{targetContent.description}</p>
+                {hasSteps && <div className="teaching-bubble-footer">
+                    {stepNumber && totalSteps && <div className="teaching-bubble-steps">
+                        {stepNumber} of {totalSteps}
+                    </div>}
+                    <div className="teaching-bubble-navigation">
+                        {hasPrevious && <Button
+                            className="secondary-button"
+                            onClick={onBack}
+                            title={backLabel}
+                            ariaLabel={backLabel}
+                            label={backLabel}
+                        />}
+                        {hasNext && <Button
+                            className="primary-button"
+                            onClick={onNext}
+                            title={nextLabel}
+                            ariaLabel={nextLabel}
+                            label={nextLabel}
+                        />}
+                        {!hasNext && <Button
+                            className="primary-button"
+                            onClick={onClose}
+                            title={finishLabel}
+                            ariaLabel={finishLabel}
+                            label={finishLabel}
+                        />}
+                    </div>
+                </div>}
+                {(!hasSteps) && <div className="teaching-bubble-footer">
```

Comment:
nit; this looks pretty similar to the previous `teaching-bubble-footer`; if we made `teaching-bubble-steps` not show up when `!hasSteps` could we avoid splitting this into 2 footer paths (can always do `const finishLabel = hasSteps ? lf("Finish") : lf("Got it")` to make the label match what you want)?

It's not a big deal to split them up if we have to, but it is an extra place to forget to update something / have to test / reason about when we come back to fix something in the future.

```tsx
{stepNumber && hasSteps && <div className="teaching-bubble-steps">
```

Reply:
oh yea. This is much cleaner. Thanks!